### ISSUE TITLE
Be able to select arbitrary time period to view graph

### ISSUE THREAD

**Issue by lucasponce** [kiali/kiali#1421]
"As a SWS user I want to be able for my graph to display information about the service graph from an arbitrary time period"

On the service graph page we will want to provide more advanced options for selecting time spans.

JIRA:
https://issues.jboss.org/browse/KIALI-222

**Comment by jshaughn** [comment]
When discussing https://github.com/kiali/kiali/issues/1859 I entered the following [edited] comment, which is also relevant here.  I was trying to understand that issue's scope and how it fit in with this feature request.  The answer is that 1859 is an in-chart manipulation that let's you customize the time range the chart is displaying, while "freezing" refresh to allow you to observe that custom range. 

Unlike the chart, the graph doesn't [currently] have a built-in/on-screen way to manipulate a time window.  That is something we'd need to solve.  And note that setting a custom time window could (probably should) be global, with a global indication of where you are in time.

"...I think we need to be careful about terminology. The graph and overview share the same duration, one minute by default. Whereas the charts (inbound/outbound traffic on a detail page) defaults to a 10 minute duration. We offer a fixed set of duration choices. Given a selected duration we query for data in the time range [(now-duration), now]. We don't allow a custom duration nor do we allow a custom endTime for the time range. The endTime is always 'now'. We also offer a fixed set of refreshInterval choices. Then, given a selected refreshInterval we slide the time range forward.

... if the refreshInterval is honored then we have created the ability to perform "replay". Note that replay could also offer SlowMotion (SM) or FastForward (FF) as we are no longer dealing with real time.

For the graph I would envision full replay ability. In this mode we would either disable, or make optional, features that query current config.

... let the user define a custom time window by setting a start and [optional] end and let Kiali run in replay mode, which would affect the graph and the charts because instead of always setting our query endTime to 'now', we'd implement a little time simulator that would return an endTime based on the defined window and refreshInterval."

In short, by letting a user define a custom start time Kiali could roll forward just like it does today, allowing users to observe/replay a past time frame of interest, or run behavioral comparisons.

**Comment by lucasponce** [comment]
Having a "Replay from" feature where an user can select from/to and perhaps Kiali can popup a "video modal" to run an animated graph of the scenario (past 2 days, hour by hour) is a very interesting idea.

I agree that we need to rely only on telemetry and we need to cut other features that are not timeseries based (configuration, also probably some workloads info, this needs to be researched).

But in short, it can provide a "what happened" feature that will help to explain things.

So +1 for this.

I propose to have some kind of modal, to have it clear that it's not the current graph, and also I remember other tooling that provided a similar feature in the same way, but that's a detail.

I like that this feature is proposed and this old issue updated.

Thanks @jshaughn.

**Comment by jotak** [comment]
Just a clarification about #1859 : it's not an in-chart control, it would rather be a modification in the existing duration component (so that would be re-usable for graph). We _also_ talk about in-chart control, but it's a following task #1860 .
Apart from that, +1 for the replay.
(and it's funny to see a feature request starting with "As a SWS user" today :-) )

**Comment by lucasponce** [comment]
> (and it's funny to see a feature request starting with "As a SWS user" today :-) )

Yes, my comments on #1859 where also related to move forward of these old filed JIRAs, where we have a lot of interesting ideas.
